Move to Quark

capital-G opened the pull request:
Move the library to a quark as discussed in #1 

Currently the original functions are omitted here but mainly because I don't know where the proper space is for such in a quark.

Todo:

* [x] Multi-channel expansion does not work.
* [x] Currently setting a manual trigger does not work, e.g. `ThatFreqAnalyzer(\b, {|f| f.postln}, Impulse.kr(1.0));`
* [x] Clarification of That vs. ThatDef
* [x] How does one modify the parameters of an analyzer, such as e.g. threshold
* [x] Write docs
* [x] Write tests
* [x] Gear this towards machine listening/analyzing or more generally as a scsynth/ugen to sclang back-communication via OSC?
* [x] Implement all Analyzers from the function version

capital-G committed:
first draft of quark

capital-G committed:
remove .DS_Store files

capital-G committed:
remove ThatInput

capital-G committed:
define trigger via lambda function

capital-G committed:
remove debug postln

capital-G committed:
modify multichannel behaviour

capital-G committed:
harden interface

capital-G committed:
rename v to latestValue

capital-G committed:
fix def behaviour

capital-G committed:
separating analyzerFunction and running analyzer

capital-G committed:
fix setting analyzerFunction in def notation

capital-G committed:
ensure that callback is a function and not nil

capital-G committed:
replace subclasses with constructors

capital-G committed:
add freqTime constructor

capital-G committed:
add identity constructor

capital-G committed:
refactoring triggerFunction

capital-G committed:
add mfcc constructor

capital-G committed:
start documentation

telephon commented on .gitignore:
```diff
@@ -0,0 +1,26 @@
+# General
```
gitignore should not be necessary in this folder.

capital-G replied:
For me it avoids committing the `.DS_Store` which macOS automatically generates. How do  you handle this stuff or do you omit `git add .`?

telephon replied:
I use a global ignore file for that `~/.gitignore_global`.
For me, it looks like this:
```
*~
.DS_Store
```

In general, there are no ignore files in the repo, because it is platform and user independent.

capital-G replied:
I think the main-benefit of providing a `.gitignore` is that it is easier for new people to contribute b/c they don't commit unnecessary files. SC does the same as well
https://github.com/supercollider/supercollider/blob/0ff1fe45db3dbab937596f42d678112eff104399/.gitignore#L2

But we can delete this file if you insist on it.

telephon replied:
>SC does the same as well
https://github.com/supercollider/supercollider/blob/0ff1fe45db3dbab937596f42d678112eff104399/.gitignore#L2

Ah yes, I didn't notice. Let's reuse that one then, ok?

telephon commented on HelpSource/Classes/That.schelp:
```diff
@@ -0,0 +1,337 @@
+For more information on this please refer to link::Guides/OSC_communication::.
+
+But communicating values from the server (so a value of a link::Classes/UGen::) to sclang has always been a bit tricky but possible via link::Classes/SendReply::.
+That is a framework which helps reducing the overhead of the server->language communication.
```
cognitive overhead 
or:
maintenance overhead

telephon commented on classes/that.sc:
```diff
@@ -0,0 +1,254 @@
+			numInputChannels = inputChannels.size;
+
+			analyzerResults = inputChannels.collect { |inputChannel, i|
+				// is .value(...) some multichannel stuff?
```
>// is .value(...) some multichannel stuff?
sorry, I don't understand

capital-G replied:
`.value` is a function call on the signal, but whats with the arguments? Is there some documentation on this call/function?

telephon replied:
`value` is a call not on a signal, but on a function that returns an event with signal graphs of UGens. If it specifies the arguments (which is optional), it can make the resulting signal graphs depend on them or add extra information to the event. E.g. besides each channel, it can access all other input channels and the index of the current channel (in a multichannel analysis signal).

capital-G replied:
Thanks for the information, wasn't aware that this is a thing in SC. added the exact phrase as a comment for future lurkers or contributors, I hope this is ok for you.

telephon commented on classes/that.sc:
```diff
@@ -0,0 +1,254 @@
+
+	*prCreateTrigger {|in, defaultTrigger, triggerFunction|
+		^if(triggerFunction.isNil, {
+			defaultTrigger;
```
better no semicolon for return values (just stylistic)

telephon replied:
(this is the same in all cases above and below)

capital-G replied:
thanks for the hint, was not aware of this style convention

telephon commented on that.quark:
```diff
@@ -0,0 +1,11 @@
+	\path: 			"that",
+	\summary:		"",
+	\isCompatible:	{ Main.versionAtMost(3, 10).not },
+	\author: 		"Julian Rohrhuber",
```
please add yourself.

telephon reviewed (changes requested):
Not yet ready, but good. The interface needs discussion:
These analysis methods are very special and often need to be adjusted for specific usage. This is why I made it all user programmable. The current version suggests that it is the final word.

Please don't remove the function library, we still want to have something like that. We could organise it into two folders.

capital-G committed:
add custom analyzer function

capital-G commented:
Analyzer function: I extended the docs so they include an example on how to use custom analyzer functions. is this sufficient or how could this be promoted in a better way?

Functions: What would be the best way to include these? Should they be loaded/imported as well or should they be accessible via docs/scd document? I am not familiar with best practices concerning functions in a SC library.

telephon commented on HelpSource/Classes/That.schelp:
```diff
@@ -0,0 +1,373 @@
+SUBSECTION::Technical
+
+SuperCollider is designed in such a way that sclang sends OSC messages to the server.
+This way of communication is well established, e.g. when adding a link::Classes/SynthDef:: to the server it gets transferred from sclang to the server via OSC as the server is only capable of generatin sound.
```
generatin –> generating

telephon reviewed (commented):
My main point is that I find it misleading that some part of the interface are methods and some are "custom" functions. The methods are just as arbitrary, they are actually just prototypes to start rolling your own.

Architecturally, I'd like to propose a different approach: you define _all_ analysis functions like this:
`ThatDef(\amp, { … })`

And then use it either by a different class:

`ThatAnalyse(\x, \amp, { ... }, { ... } )`

what do you think?

capital-G commented:
IMO the problem is that people in sclang do not use a lot of additional constructors (https://kotlinlang.org/docs/classes.html#secondary-constructors and  https://stackoverflow.com/a/682545) - but this case would be an excellent example for an constructor: It is possible to create something completely on your own form scratch but also to use "presets" which cover the most common usages and provide examples via source code. Yet they all get created in the same namespace and behave exactly the same.

I abandoned the Multi-Class approach b/c I could not find an intuitive way to make it clear to the user that all That instances share the same namespace and therefore clashing is vorprogrammiert.

```supercollider
ThatAmp(\foo, ...)

ThatFreq(\foo, ...)
// i would not understand that ThatFreq overwrites now ThatAmp
```

Adding a prefix for each instance could fix this but then it is not intuitive what this prefix actually is.

```supercollider
ThatAmp(\foo)

That(\foo)
// no

That(\amp_foo)
// how do i know it is amp_ ?

// or one accepts that there is no possible way to access all That instances but I find this limiting
```

telephon commented on classes/that.sc:
```diff
@@ -0,0 +1,254 @@
+	}
+
+	prCreateOscDef {
+		oscdef = OSCdef(defName, { |msg|
```
better use an `OSCFunc` unless you need the OSCdef anywhere else.

capital-G committed:
change gitignore

capital-G committed:
add comment on .value call

capital-G committed:
remove semicolons on return statements

capital-G committed:
add another author

capital-G committed:
fix typo

capital-G committed:
specifiy kind of maintenance in docs

capital-G committed:
change from OSCdef to OSCFunc

capital-G commented:
For my part we are done except for Docs and Tests

**Regarding documentation**: we should indeed change the framing more towards its default to write its own analyzerFunctions and not call them `custom` functions. This can be *manifestiert* by demonstrating this behavior directly in the introduction example.
Although the constructors provide a certain way of non-complete pre-determination for the use of That I am actually fine with this as I personally use it mostly for envelope and pitch tracking and I think for everything beyond that the people are already attached enough so they will read the docs an how to expand the usage of That.

**Regarding tests**: I think a simple test of callbacks, latest value and garbage collection should be sufficient.

telephon commented:
>For my part we are done except for Docs and Tests

OK, so let us take the challenge and use these constructors. We should note that they are work in progress, they need to be improved. Versioning might be good once we are a bit more stable.


Two issues are left: 
- can we pass in signal functions instead of Ndefs as we did in the functionasl version?
- can we please leave the older functional version in the repo?

capital-G commented:
> > For my part we are done except for Docs and Tests
> 
> OK, so let us take the challenge and use these constructors. We should note that they are work in progress, they need to be improved. Versioning might be good once we are a bit more stable.

I think the constructors should be more or less trivial and more seen as convenience and therefore I see not that much of a room to be improved for e.g. the amp envelop follower, but I am always pro to keep things in motion and allow breaking changes as long as those are documented.
Do you have anything in mind that could change?

> Two issues are left:
> 
> * [x] can we pass in signal functions instead of Ndefs as we did in the functionasl version?

Yes - will add this in the docs as an example as well.
But maybe we need additional housekeeping for this. What should happen with the function once we clear the `That` instance? If we do not remove it it will be a silent synth on the server as there is no handle to this or if we do remove it we may introduce unintended behaviour (why did stopping the analysis also stopped my sound?)

>  * [ ] can we please leave the older functional version in the repo?

Totally, but I do not know what is the best way to include it? Via docs? To copy and paste it from the git repo? Or as static functions? (which could create some maintenance overhead in the long run?)

capital-G committed:
Revert "change from OSCdef to OSCFunc"

This reverts commit 11acbf3135f4fde876238d28893d01ac04c03fcd as OSCdef
is more fitting when we update the OSC callback function.

capital-G commented:
I reverted the change from OSCFunc to OSCdef as this introduced a bug when one updates the callback function. Instead of taking care of this manually (although it would be easy as we just have to check if oscFunc is already defined) I think we can make this more simple by using OSCdef.

capital-G committed:
add more spotlight on non-constructor usage

capital-G committed:
add markov example

capital-G committed:
add unit tests

capital-G committed:
update readme

telephon commented:
> Do you have anything in mind that could change?

Not yet, but I know that I haven't fine tuned these things. I am completely in agreement about just making changes for a while.


> But maybe we need additional housekeeping for this. What should happen with the function once we clear the `That` instance?

~~You create an internal `NodeProxy`, and when you clear the `That` instance you also call `clear` on the `NodeProxy`. The same should be done for the `OSCFunc`. Using `OSCdef` is kind of OK, but may break peoples' code, when they have another instance of it which happens to have the same name as the `That`.~~

I just saw that we render the special name with `defName = "that_%".format(name).asSymbol;`. This should be fine for now.

The function gets thrown away and you `clear` the internal Ndef that holds it. I think all that already works, just the examples use `Ndef` as an input and happen to work, because `Ndef` returns a UGen graph on calling `value`.

> I reverted the change from OSCFunc to OSCdef as this introduced a bug when one updates the callback function.

Did you perhaps find a bug in `OSCFunc` when setting its function? If yes, we should use an `OSCFunc`, and just throw it away once we want a different function.



> > * can we please leave the older functional version in the repo?
> 
> Totally, but I do not know what is the best way to include it?

There is no overhead, only the folder `Classes/` is searched by sclang class library compilation. We just put it in a different folder, let's say called `/That/that-environment/`.

telephon commented on README.md:
````diff
@@ -0,0 +1,39 @@
+Quarks.install("https://github.com/musikinformatik/that.git");
+```
+
+After installation please re-compile the Class Library (aka restart the
````
aka –> or (not the same thing)

telephon commented on classes/that.sc:
```diff
@@ -0,0 +1,329 @@
+				Error("Please provide an analyzerFunction").throw;
+			});
+			callback = callback ? {}; // enforce a function as callback and not allow nil
+			res = super.newCopyArgs(name, input, analyzerFunction, callback).init.();
```
`.init`

`.init.()` is the same as `.init.value` which just happens to be a no-op

telephon commented on classes/that.sc:
```diff
@@ -0,0 +1,329 @@
+		}, oscChannelName).fix;
+	}
+
+	*prCreateTrigger {|in, defaultTrigger, triggerFunction|
```
This method is a bit opaque. We may have to change it, but fine for now.

capital-G replied:
Its more of a convenience method caused by the [rule of 3](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming))

telephon reviewed (changes requested):
Just some minor things.
And let's keep the that-system-function adn that-intro files.

capital-G committed:
state installation more precisely

capital-G committed:
change init call

capital-G committed:
reintroduce original version

capital-G commented:
> Did you perhaps find a bug in `OSCFunc` when setting its function? If yes, we should use an `OSCFunc`, and just throw it away once we want a different function.

I switched back to OSCdef b/c on OSCFunc I have to take care not just of the updating of the function but also on releasing the old. It is 2 lines to check but it introduces complexity where there is a mechanism like OSCdef which can take care of this for us. The only advantage would be to have a more private, encapsulated OSC function which is nice from an engineering perspective, but then we would really need more tests for this behavior and I am really lazy at writing tests so I try to write as few code as reasonable possible.

---

I added the old version again, I hope I did not destroy the git history too much doing it.
